# Finding _Rx_ in Potato - an example workflow

Example output files are provided in the example_outputs directory. These can be compared to the files generated by this workflow (eg. by checksums) to confirm correct functionality.

## Download required files

All files, except for the reference DM potato genome are available in this repository, simply download the latest release to your local machine. All commands in this workflow will assume you are based in this repository, you may need to change commands if you change the structure.

Ensure you have followed the instructions in README.md to install required software and optionally set up a cluster profile.

```bash
# Fetch the DM genome, this is too large to include here
cd example_inputs/agrenseq
wget http://spuddb.uga.edu/data/PGSC_DM_v4.03_pseudomolecules.fasta.zip
unzip PGSC_DM_v4.03_pseudomolecules.fasta.zip
```

## Obtain reads

Reads have been submitted to ENA BioProjects PRJEB56823 and PRJEB56825. In this example we will use the GNU parallel utility to download reads. Replace X in the parallel command with the number of jobs to run in parallel, this will depend on your system, based on CPU availability and networking bandwidth. If you prefer another method you may use this, though you may need to change file paths in later files.

```bash
mkdir -p ../Reads
cd ../Reads
parallel -j X wget < ../reads.txt
```

## Perform SMRT-RenSeq assembly

Ensure to add the absolute path to the pacbio reads (ERR10385989) to your sample.txt example input.

```bash
mkdir -p /path/to/directory/for/assembly
cd /path/to/directory/for/assembly
nextflow run Hutton-Potato-Genetics/nfHiss --workflow smrtrenseq --reads /path/to/example_inputs/smrtrenseq/sample.txt --genome_size 2000000 --max_input_coverage 20000 --flanking 1000 --five_prime GGTAGT --three_prime ACTACC
```

## Perform AgRenSeq

Ensure to add the paths to your downloaded reads to read_scores.txt

```bash
mkdir -p /path/to/directory/for/association
cd /path/to/directory/for/association
nextflow run Hutton-Potato-Genetics/nfHiss --workflow agrenseq --reads /path/to/example_inputs/agrenseq/read_scores.txt --adaptor_1 /path/to/example_inputs/agrenseq/Index2varN.fasta --adaptor_2  /path/to/example_inputs/agrenseq/Universal_var.fasta --association_reference /path/to/assembly/from/smrtrenseq --blast_reference /path/to/PGSC_DM_v4.03_pseudomolecules.fasta --threshold 26 --title Gemson --annotator_bed /path/to/nlr/bed/from/smrtrenseq
```

## Perform drenseq

Ensure to add the paths to your downloaded reads to samples.tsv

```bash
mkdir -p /path/to/directory/for/drenseq
cd /path/to/directory/for/drenseq
nextflow run Hutton-Potato-Genetics/nfHISS --workflow drenseq --reference /path/to/example_inputs/drenseq/Gemson_candidates.fa --bed /path/to/example_inputs/drenseq/Gemson_candidates.bed --reads /path/to/example_inputs/drenseq/samples.tsv --score "L,0,-0.24" --max_align 10 --baits /path/to/example_inputs/drenseq/Solanum_baits.fasta --identity 90 --coverage 90 --flank 0 --ulimit 1024 --adaptor_1 /path/to/example_inputs/drenseq/Index2varN.fasta --adaptor_2 /path/to/example_inputs/drenseq/Universal_var.fasta
```

This should show tig00001343\_nlr_1 as the only candidate with 100% coverage in all samples scored as 1 in the read scores file used for AgRenSeq. The nucleotide sequence for this can be found in the NLR Annotator output fasta in the SMRT-RenSeq assembly directory. When BLASTed against the NCBI nr/nt database, the top hit is the reference _Rx_ sequence.
